In [188]:
# Import the necessary libraries
import pandas as pd
from IPython.display import display

In [189]:
train = pd.read_csv('Train.csv')
test = pd.read_csv('Test.csv')
description = pd.read_csv('VariableDefinitions.csv')

## 2. Data Understanding
- Data info
- Data shape
- Missing data
- Duplicated data

Description of the data in use

In [190]:
description

Variable Definitions                                         Unnamed: 1
0                  country                         Country interviewee is in.
1                     year                           Year survey was done in.
2                 uniqueid             Unique identifier for each interviewee
3            location_type                     Type of location: Rural, Urban
4         cellphone_access  If interviewee has access to a cellphone: Yes, No
5           household_size               Number of people living in one house
6        age_of_respondent                         The age of the interviewee
7     gender_of_respondent                Gender of interviewee: Male, Female
8   relationship_with_head  The interviewee’s relationship with the head o...
9           marital_status  The martial status of the interviewee: Married...
10         education_level  Highest level of education: No formal educatio...
11                job_type  Type of job interviewee has: Farming and Fishi...

In [191]:
class DataUnderstanding:
	'''
	This is a class that will handle all the data understanding process;
	The data info, shape of the data, checking for missing values, 
	checking for duplications in the data and data overview
	'''
	def __init__(self,df):
		self.df = df
		self.shape = self.df.shape
		self.info = self.df.info
		self.describe = self.df.describe
		self.head = self.df.head(5)
		self.duplicates = self.df.duplicated().sum()
		self.missing = self.df.isna().sum()

In [192]:
# Initialize the data understanding class
train_df = DataUnderstanding(train)
test_df = DataUnderstanding(test)

Data Overview

In [193]:
train_df.head

country  year    uniqueid bank_account location_type cellphone_access  \
0   Kenya  2018  uniqueid_1          Yes         Rural              Yes   
1   Kenya  2018  uniqueid_2           No         Rural               No   
2   Kenya  2018  uniqueid_3          Yes         Urban              Yes   
3   Kenya  2018  uniqueid_4           No         Rural              Yes   
4   Kenya  2018  uniqueid_5           No         Urban               No   

   household_size  age_of_respondent gender_of_respondent  \
0               3                 24               Female   
1               5                 70               Female   
2               5                 26                 Male   
3               5                 34               Female   
4               8                 26                 Male   

  relationship_with_head           marital_status  \
0                 Spouse  Married/Living together   
1      Head of Household                  Widowed   
2         Other relative     Single/Never Married   
3      Head of Household  Married/Living together   
4                  Child     Single/Never Married   

                   education_level                   job_type  
0              Secondary education              Self employed  
1              No formal education       Government Dependent  
2  Vocational/Specialised training              Self employed  
3                Primary education  Formally employed Private  
4                Primary education        Informally employed

In [194]:
test_df.head

country  year       uniqueid location_type cellphone_access  household_size  \
0   Kenya  2018  uniqueid_6056         Urban              Yes               3   
1   Kenya  2018  uniqueid_6060         Urban              Yes               7   
2   Kenya  2018  uniqueid_6065         Rural               No               3   
3   Kenya  2018  uniqueid_6072         Rural               No               6   
4   Kenya  2018  uniqueid_6073         Urban               No               3   

   age_of_respondent gender_of_respondent relationship_with_head  \
0                 30                 Male      Head of Household   
1                 51                 Male      Head of Household   
2                 77               Female                 Parent   
3                 39               Female      Head of Household   
4                 16                 Male                  Child   

            marital_status                  education_level  \
0  Married/Living together              Secondary education   
1  Married/Living together  Vocational/Specialised training   
2  Married/Living together              No formal education   
3  Married/Living together                Primary education   
4     Single/Never Married              Secondary education   

                       job_type  
0  Formally employed Government  
1     Formally employed Private  
2          Remittance Dependent  
3          Remittance Dependent  
4          Remittance Dependent

In [195]:
# Shape of the data
print(f'''
Shape of train df is {train_df.shape} with {train_df.shape[0]} rows and {train_df.shape[1]} columns
Shape of test df is {test_df.shape} with {test_df.shape[0]} rows and {test_df.shape[1]} columns
''')


Shape of train df is (23524, 13) with 23524 rows and 13 columns
Shape of test df is (10086, 12) with 10086 rows and 12 columns



General Information about the data

In [196]:
train_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23524 entries, 0 to 23523
Data columns (total 13 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   country                 23524 non-null  object
 1   year                    23524 non-null  int64 
 2   uniqueid                23524 non-null  object
 3   bank_account            23524 non-null  object
 4   location_type           23524 non-null  object
 5   cellphone_access        23524 non-null  object
 6   household_size          23524 non-null  int64 
 7   age_of_respondent       23524 non-null  int64 
 8   gender_of_respondent    23524 non-null  object
 9   relationship_with_head  23524 non-null  object
 10  marital_status          23524 non-null  object
 11  education_level         23524 non-null  object
 12  job_type                23524 non-null  object
dtypes: int64(3), object(10)
memory usage: 2.3+ MB


In [197]:
test_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10086 entries, 0 to 10085
Data columns (total 12 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   country                 10086 non-null  object
 1   year                    10086 non-null  int64 
 2   uniqueid                10086 non-null  object
 3   location_type           10086 non-null  object
 4   cellphone_access        10086 non-null  object
 5   household_size          10086 non-null  int64 
 6   age_of_respondent       10086 non-null  int64 
 7   gender_of_respondent    10086 non-null  object
 8   relationship_with_head  10086 non-null  object
 9   marital_status          10086 non-null  object
 10  education_level         10086 non-null  object
 11  job_type                10086 non-null  object
dtypes: int64(3), object(9)
memory usage: 945.7+ KB


Check for missing data

In [198]:
train_df.missing

country                   0
year                      0
uniqueid                  0
bank_account              0
location_type             0
cellphone_access          0
household_size            0
age_of_respondent         0
gender_of_respondent      0
relationship_with_head    0
marital_status            0
education_level           0
job_type                  0
dtype: int64

In [199]:
test_df.missing

country                   0
year                      0
uniqueid                  0
location_type             0
cellphone_access          0
household_size            0
age_of_respondent         0
gender_of_respondent      0
relationship_with_head    0
marital_status            0
education_level           0
job_type                  0
dtype: int64

Check for duplicates in the data

In [200]:
print(f'''
Train df: There are {train_df.duplicates} duplicated records in the dataframe
Test df: There are {test_df.duplicates} duplicated records in the dataframe 
''')


Train df: There are 0 duplicated records in the dataframe
Test df: There are 0 duplicated records in the dataframe 



Summary Statistics

In [201]:
train_df.describe()

year  household_size  age_of_respondent
count  23524.000000    23524.000000       23524.000000
mean    2016.975939        3.797483          38.805220
std        0.847371        2.227613          16.520569
min     2016.000000        1.000000          16.000000
25%     2016.000000        2.000000          26.000000
50%     2017.000000        3.000000          35.000000
75%     2018.000000        5.000000          49.000000
max     2018.000000       21.000000         100.000000

In [202]:
test_df.describe()

year  household_size  age_of_respondent
count  10086.000000    10086.000000       10086.000000
mean    2016.975907        3.778802          38.308348
std        0.847356        2.212721          16.270053
min     2016.000000        1.000000          16.000000
25%     2016.000000        2.000000          26.000000
50%     2017.000000        3.000000          35.000000
75%     2018.000000        5.000000          48.000000
max     2018.000000       20.000000         100.000000

Summary of the Data Understanding:
- The train data:
    - There are no missing or duplicated records in the data.
    - There are 3 numerical and 9 categorical columns
    - It consists of 23524 rows and 13 columns
- The test data
    - There are no missing or duplicated records in the data.
    - There are 3 numerical and 9 categorical columns
    - It consists of 10086 rows and 12 columns

## 3. Data Preparation
- Convert data types -> object to categorical 
- Clean the data
    - Check for uniqueness of the data -> Combine uniqueid and counry to form a truly unique field
- Selecting the necesary columns
- Perform EDA analysis
    - Univariate analysis
    - Bivariate analysis
- Data Preprocessing
    - Encoding
    - Standardization? or normalization 

In [203]:
data = train_df.df
data.head()

country  year    uniqueid bank_account location_type cellphone_access  \
0   Kenya  2018  uniqueid_1          Yes         Rural              Yes   
1   Kenya  2018  uniqueid_2           No         Rural               No   
2   Kenya  2018  uniqueid_3          Yes         Urban              Yes   
3   Kenya  2018  uniqueid_4           No         Rural              Yes   
4   Kenya  2018  uniqueid_5           No         Urban               No   

   household_size  age_of_respondent gender_of_respondent  \
0               3                 24               Female   
1               5                 70               Female   
2               5                 26                 Male   
3               5                 34               Female   
4               8                 26                 Male   

  relationship_with_head           marital_status  \
0                 Spouse  Married/Living together   
1      Head of Household                  Widowed   
2         Other relative     Single/Never Married   
3      Head of Household  Married/Living together   
4                  Child     Single/Never Married   

                   education_level                   job_type  
0              Secondary education              Self employed  
1              No formal education       Government Dependent  
2  Vocational/Specialised training              Self employed  
3                Primary education  Formally employed Private  
4                Primary education        Informally employed

In [204]:
# Check for uniqueness for uniqueid column
print(f'Number of unique rows in uniqueid {data["uniqueid"].nunique()} out of {len(data)}')

Number of unique rows in uniqueid 8735 out of 23524


In [205]:
# Group by unique id to observe what is going on
data.groupby(by=['uniqueid','country']).first().head(10)

year bank_account location_type cellphone_access  \
uniqueid     country                                                      
uniqueid_1   Kenya     2018          Yes         Rural              Yes   
             Rwanda    2016           No         Urban              Yes   
             Tanzania  2017           No         Rural              Yes   
             Uganda    2018           No         Rural               No   
uniqueid_10  Kenya     2018           No         Urban              Yes   
             Rwanda    2016          Yes         Urban              Yes   
             Tanzania  2017           No         Urban               No   
             Uganda    2018           No         Rural               No   
uniqueid_100 Kenya     2018          Yes         Urban              Yes   
             Rwanda    2016           No         Rural              Yes   

                       household_size  age_of_respondent gender_of_respondent  \
uniqueid     country                                                            
uniqueid_1   Kenya                  3                 24               Female   
             Rwanda                 7                 18               Female   
             Tanzania               6                 22               Female   
             Uganda                 5                 32               Female   
uniqueid_10  Kenya                  3                 76               Female   
             Rwanda                 4                 28                 Male   
             Tanzania               2                 65               Female   
             Uganda                 8                 45               Female   
uniqueid_100 Kenya                  5                 42                 Male   
             Rwanda                 5                 31               Female   

                      relationship_with_head           marital_status  \
uniqueid     country                                                    
uniqueid_1   Kenya                    Spouse  Married/Living together   
             Rwanda                    Child     Single/Never Married   
             Tanzania                  Child       Divorced/Seperated   
             Uganda        Head of Household       Divorced/Seperated   
uniqueid_10  Kenya         Head of Household       Divorced/Seperated   
             Rwanda                    Child     Single/Never Married   
             Tanzania                 Spouse     Single/Never Married   
             Uganda                    Child  Married/Living together   
uniqueid_100 Kenya         Head of Household  Married/Living together   
             Rwanda                   Spouse  Married/Living together   

                           education_level                      job_type  
uniqueid     country                                                      
uniqueid_1   Kenya     Secondary education                 Self employed  
             Rwanda    Secondary education          Remittance Dependent  
             Tanzania  Secondary education                     No Income  
             Uganda      Primary education                  Other Income  
uniqueid_10  Kenya     No formal education          Remittance Dependent  
             Rwanda     Tertiary education  Formally employed Government  
             Tanzania    Primary education                 Self employed  
             Uganda    No formal education                 Self employed  
uniqueid_100 Kenya     Secondary education           Farming and Fishing  
             Rwanda      Primary education           Farming and Fishing

We can see that for every country there is a similar uniqueid. We can combine the country and the uniqueid to form a completly uniqueid

In [211]:
# Combine the columns and drop country
data['uniqueid'] = data[['uniqueid', 'country']].apply(lambda x: ' x '.join(x.values),axis=1)
data.head()

country  year                    uniqueid bank_account location_type  \
0   Kenya  2018  uniqueid_1 x Kenya x Kenya          Yes         Rural   
1   Kenya  2018  uniqueid_2 x Kenya x Kenya           No         Rural   
2   Kenya  2018  uniqueid_3 x Kenya x Kenya          Yes         Urban   
3   Kenya  2018  uniqueid_4 x Kenya x Kenya           No         Rural   
4   Kenya  2018  uniqueid_5 x Kenya x Kenya           No         Urban   

  cellphone_access  household_size  age_of_respondent gender_of_respondent  \
0              Yes               3                 24               Female   
1               No               5                 70               Female   
2              Yes               5                 26                 Male   
3              Yes               5                 34               Female   
4               No               8                 26                 Male   

  relationship_with_head           marital_status  \
0                 Spouse  Married/Living together   
1      Head of Household                  Widowed   
2         Other relative     Single/Never Married   
3      Head of Household  Married/Living together   
4                  Child     Single/Never Married   

                   education_level                   job_type  
0              Secondary education              Self employed  
1              No formal education       Government Dependent  
2  Vocational/Specialised training              Self employed  
3                Primary education  Formally employed Private  
4                Primary education        Informally employed

In [212]:
# Confirm the uniqueness for uniqueid column
print(f'Number of unique rows in uniqueid {data["uniqueid"].nunique()} out of {len(data)}')

Number of unique rows in uniqueid 23524 out of 23524


In [216]:
# Select the columns we will use
df = data.copy()
df.drop(['uniqueid'],axis=1,inplace=True)
df.head()

country  year bank_account location_type cellphone_access  household_size  \
0   Kenya  2018          Yes         Rural              Yes               3   
1   Kenya  2018           No         Rural               No               5   
2   Kenya  2018          Yes         Urban              Yes               5   
3   Kenya  2018           No         Rural              Yes               5   
4   Kenya  2018           No         Urban               No               8   

   age_of_respondent gender_of_respondent relationship_with_head  \
0                 24               Female                 Spouse   
1                 70               Female      Head of Household   
2                 26                 Male         Other relative   
3                 34               Female      Head of Household   
4                 26                 Male                  Child   

            marital_status                  education_level  \
0  Married/Living together              Secondary education   
1                  Widowed              No formal education   
2     Single/Never Married  Vocational/Specialised training   
3  Married/Living together                Primary education   
4     Single/Never Married                Primary education   

                    job_type  
0              Self employed  
1       Government Dependent  
2              Self employed  
3  Formally employed Private  
4        Informally employed

In [221]:
# Show categorical and numerical 
categorical =  df.columns[df.dtypes == "object"].tolist()
numerical = df.columns[df.dtypes == "int64"].tolist()
print("\033[35m" + 'Categorical columns:' + "\033[0m")
display(categorical)
print("\033[35m" + 'Numerical columns:' + "\033[0m")
display(numerical)


Categorical columns:


['country',
 'bank_account',
 'location_type',
 'cellphone_access',
 'gender_of_respondent',
 'relationship_with_head',
 'marital_status',
 'education_level',
 'job_type']

Numerical columns:


['year', 'household_size', 'age_of_respondent']

In [222]:
# Convert the object data to category data
for column in categorical:
    data[column] = data[column].astype('category')
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23524 entries, 0 to 23523
Data columns (total 13 columns):
 #   Column                  Non-Null Count  Dtype   
---  ------                  --------------  -----   
 0   country                 23524 non-null  category
 1   year                    23524 non-null  int64   
 2   uniqueid                23524 non-null  object  
 3   bank_account            23524 non-null  category
 4   location_type           23524 non-null  category
 5   cellphone_access        23524 non-null  category
 6   household_size          23524 non-null  int64   
 7   age_of_respondent       23524 non-null  int64   
 8   gender_of_respondent    23524 non-null  category
 9   relationship_with_head  23524 non-null  category
 10  marital_status          23524 non-null  category
 11  education_level         23524 non-null  category
 12  job_type                23524 non-null  category
dtypes: category(9), int64(3), object(1)
memory usage: 943.7+ KB


I'm here
- Encode the categorical columns
    - LabelEncoder - Target variable
    - Others - OHE
- Perform EDA:
    - Univariate
    - Bivariate
    - Please don't use pie charts
    - Maybe use pivot tables like Ken Jee?

In [ ]:
# Encode the categorical columns


In [ ]:
class DataPreparation:
    # 
    

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23524 entries, 0 to 23523
Data columns (total 13 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   country                 23524 non-null  object
 1   year                    23524 non-null  int64 
 2   uniqueid                23524 non-null  object
 3   bank_account            23524 non-null  object
 4   location_type           23524 non-null  object
 5   cellphone_access        23524 non-null  object
 6   household_size          23524 non-null  int64 
 7   age_of_respondent       23524 non-null  int64 
 8   gender_of_respondent    23524 non-null  object
 9   relationship_with_head  23524 non-null  object
 10  marital_status          23524 non-null  object
 11  education_level         23524 non-null  object
 12  job_type                23524 non-null  object
dtypes: int64(3), object(10)
memory usage: 2.3+ MB
